Sources/Bond/ObservableTree/TreeNodeProtocol.swift

```diff
@@ -0,0 +1,57 @@
+//
+//  The MIT License (MIT)
+//
+//  Copyright (c) 2018 DeclarativeHub/Bond
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+public protocol TreeNodeProtocol: MutableCollection where Index == IndexPath, Element == Self {
+
+    associatedtype Value
+    var value: Value { get }
+    var children: [Self] { get }
```

Comment:
Is there a way to express this without forcing it to be an array? Making this Collection or MutableCollection would solve some headaches for me.

Reply:
It'd also be nice to have a "root" tree node that returns `value` as `nil`, or a type that's not the same as `Element`. 

I have a structure I'm working against that is:

- Page
    - Layer
    - Layer
    - Layer
        - Layer
        - Layer
    - Layer

etc, etc.

Reply:
Yes, I think we can make children a collection.

As for the `Value` type (not to be confused with `Element` which is another `ThreeNode<Value>`), could you maybe define your tree as `TreeNode<Layer?>`? It should have the same effect. 

Alternatively, maybe a better solution would be to use enum to define all possible tree node type:

```
enum Item {
    case page(Page)
    case layer(Layer)
}
```

Reply:
How do you see the enum work in this context? That's obviously preferable, but I'm not sure how it fits alongside what we're designing here.

Reply:
Would that work for you? I don't think it's possible to make a tree with different value types on different levels, at least no without having an infinite number of generic arguments 😅